add pagination guide (#1345)
strawberry-graphql/strawberry: docs/guides/pagination/overview.md

---
title: Pagination - Overview
---

# Pagination

Whenever we deal with lists in GraphQL, we usually need to limit the number of items returned. Surely, we don't want to send massive lists of
items that take a considerable toll on the server! The goal of this guide is to help you get going fast with pagination!

## Pagination at a glance

> We have always dealt with pagination in different situations. Let us take a look at some of the common ways pagination

Reviewer: I would hesitate to use "we have always". This could be someone's first shot at pagination and they're coming in with no context.

Author: Yeah, that makes sense. I'll probably remove that!

can be implemented today!

### Offset based pagination

This pagination style is similar to the syntax we use when looking up database records. Here, the client specifies the number of result to be

> obtained at a time, along with an offset- which usually denotes the number of results to be skipped from the beginning. This type of pagination

Reviewer 2: ```suggestion
obtained at a time, along with an offset which usually denotes the number of results to be skipped from the beginning. This type of pagination
```

> This pagination style is similar to the syntax we use when looking up database records. Here, the client specifies the number of result to be
> obtained at a time, along with an offset- which usually denotes the number of results to be skipped from the beginning. This type of pagination

Reviewer: ```suggestion
This pagination style is similar to the syntax we use when looking up database records. Here, the client specifies the number of results to be
obtained at a time, along with an offset, which usually denotes the number of results to be "skipped" at the start. This type of pagination
```

> is widely used. Implementing offset-based pagination with an SQL database is straight-forward:

Reviewer: ```suggestion
is widely used. Implementing offset-based pagination with a SQL database is straight-forward:
```
This might start some holy wars :grin:

> - We count all of the results to determine the total number of pages
> - We use the limit and offset values given to query for the items in the requested page.

Reviewer: ```suggestion
- We count all of the results to determine the total number of pages
- We use the `limit` and `offset` values given to query for the items in the requested page.
```

Offset based pagination also provides us the ability to jump to a specific page in a dataset.

Let us understand offset based pagination better, with an example. Let us assume that we want to request a list of users, 2 at a time, from a server.
We start out be sending a request to the server, with the desired limit and offset values.

```json
{
  "limit": 2,
  "offset": 0
}

> ```

Reviewer: How does the request know we're asking for `users`? Do we want to add a bit more context to this snippet?

Reviewer: Actually, is this even supposed to be an actual GraphQL request? We may want to add something short that says, "Don't worry about the actual request format for now. Think about it conceptually, and we'll go into more detail on the actual implementation later"?

Author: No, this isnt graphql. It's just an intro to how pagination works. It's just a simple JSON response for conceptual understanding

Author: Yes, we could add that bit there.

The response from the server would be:

```json
{
  "users": [
    {
      "id": 1,
      "name": "Norman Osborn",
      "occupation": "Founder, Oscorp Industries",
      "age": 42
    },
    {
      "id": 2,
      "name": "Peter Parker",
      "occupation": "Freelance Photographer, The Daily Bugle",

>       "age": 16

Reviewer: The age seems a little low :thinking:

Author: Right, he was a photographer in college😅😂 I could change that

the server returns results after the given pointer. This method addresses the drawbacks of using offset pagination, but does so by making certain trade offs:

- The cursor must be based on a unique, sequential identifier in the given source.
- There is no concept of the total number of pages or results in the dataset.
- The client can’t jump to a specific page.

Let us understand cursor based pagination better, with the example given below. We want to request a list of users, 2 at a time, from
the server. We don't know the cursor initially, so we will assign it a null value.

```json
{
  "limit": 2,
  "cursor": null
}
```

The response from the server would be:

```json
{
  "users": [
    {
      "id": 3,
      "name": "Harold Osborn",
      "occupation": "President, Oscorp Industries",
      "age": 19
    },
    {
      "id": 4,
      "name": "Eddie Brock",
      "occupation": "Journalist, The Eddie Brock Report",
      "age": 20
    }
  ],
  "next_cursor": "3"
}
```

The next cursor returned by the server can be used to get the next set of users from the server.

```json
{
  "limit": 2,
  "cursor": "3"
}
```

> This is an example for forward pagination - pagination can be done backwards too!

Reviewer: ```suggestion
This is an example of forward pagination, but pagination can be done backwards too!
```
